### PR TITLE
#40 Teaser module has too much workload with large category amounts

### LINKED ISSUE #40: Teaser-Modul - DB Overload if you have many Categories
If you have much categories in Contenido, the Teaser-Module has a bottleneck in requesting those Categories and each of the concatenated Articles for "BackendView".

The Teaser-Module loads those Categories and Articles for the Popup-Dialoge

Auswahl_332

So if you have 5 Teaser at your Page, the Backend will load the function "buildCategorySelect" 10 times and have a huge overload (14509 DB-Requests if con_cat_art has 763 entries and con_cat 210 entries) of DB-Requests to preload those Categories & Articles for "EACH" of the Teasers

Auswahl_305

https://forum.contenido.org/viewtopic.php?f=98&t=39028

After those first 14500 request, another 14500 request are started for the "advanced-Tabs" of the Teaser Dialog

### PR DESCRIPTION
**Pull request by tone2tone**
If there are lots of categories and the teaser module is used multiple times, the ongoing multiple creation of category lists through buildCategoySelect causes performance issues.
To avoid that, class.content.type.teaser.php got a new function. If not present, it creates the category list and stores it in a constant. If necessary, this constant is tweaked to fit multiple use and to set values to "selected".

### CODE REVIEW

**Review by muratpurc — CHANGES_REQUESTED** [review]
@tone2tone I like the idea of caching category select to optimize the performance.

Adding the cache logic to the class cContentTypeTeaser and using a constant for it seems to be a not so good idea.

The class should be decoupled from the cache logic, it should just retrieve the data, the caching can be done in the function buildCategorySelect(), see file functions.general.php.

Constants should not be used to store cache results, a static variable would do the job, and it would not pollute the global scope.

The function buildCategorySelect() is the main place, which builds the category select, it can be used from everywhere, not only b cyContentTypeTeaser, therefore the caching should be done in this function.

Look how it is solved in the function conGetMetaValue() in functions.con2.php.

**Comment by tone2tone** [comment]
@purcaholic I strongly disagree here. buildCategorySelect can be called with parameters, e.g. preset selected categories and might return preconfigured string code. As this function is called from various places, it is NOT a good idea to hold its content there in a static variable as for the teasers we need a very specific version of this function, namely one with NO parameters, no selected categories and a replacable id that needs to be swapped afterwards. As this function can produce numerous various outcome and not "static" content, it makes no sense to cache something there as the next call would require some other outcome.

**Comment by muratpurc** [comment]
@tone2tone you can build a unique value with the parameters, use the value as a key of the static variable and the result as a value. The static variable could be a associative array.

**Comment by tone2tone** [comment]
Sorry, I am lost now. I cannot see how to achieve that within the buildCategorySelect function.

**Comment by muratpurc** [comment]
@tone2tone I pushed some changes into your branch, you can take a look at them, if you want. Not everything was related to the ticket, I also did some cleanup and small optimizations.

**Comment by tone2tone** [comment]
@purcaholic Hi Murat, I implemented your buildCategory-function, but sorry to say with 3 teasers online, it evokes the sql-statement three times, which is still unnecessary. It saves every configuration of a function call in an array, I get that, but still this requires every deviant call to be executed. I aimed for having a basic call with standard preselection (recognisable id to be replaceble, no chosen value) in order to only run the sql statement ONCE. Let me see if I can get up with some other version to reflect that based on your approach.

**Comment by muratpurc** [comment]
@tone2tone If you call the function `buildCategorySelect()` with different parameters, then it will always execute the query and create a new cache entry. This is the current logic and it may be not perfect.

Three of the function parameters are used only for the view (`$sName`, `$sValue`, `$sClass`) and only one function parameter (`$sLevel`) is used together with the globals `$client` and `$lang` to build and execute the query.

So, you can modify the logic to use the cache only to store the query results and not to store the whole generated select.

**Comment by tone2tone** [comment]
@purcaholic Will do. I will store cache in an array based on requested level, with no pre-configured options. Each requested cached version will be tweaked at the end to fit the function call parameters (replacing id, value select and class accordingly). This way, we reduce sql request to a minimum; we do not need my new function in teaser class and can call the buildCategorySelect function without changes to make it work for all existing function calls. Will submit a commit ASAP.
The buildArticle function can stay as per your proposal; this function is not really the one overloading the teaser, but still nice to reduce workload now and then.

**Comment by tone2tone** [comment]
Sorry for the commit-havac. As I said, VS Code and Git are doing their best right now to drive me crazy by doing whatever, but not what I tell them to do. Alle "merge branch" things seem to be copies of your work, only the "reworked" commit is the actual makeover.

File: contenido/includes/functions.general.php [hunk 1]
```diff
@@ -1140,66 +1135,77 @@ function buildArticleSelect($sName, $iIdCat, $sValue) {
...
+            $spaces = "&nbsp;&nbsp;" . str_repeat("&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;", $props["level"]);
+            $selectElem->appendOptionElement(new cHTMLOptionElement($spaces . ">" . $props["name"], $tmpidcat)); 
         }
+
+        $cache[$sLevel] = $selectElem->toHtml();
+
     }
 
-    return $selectElem->toHtml();
+    // now tweak standard output with desired replacements for id, value and class
```
Comment (muratpurc) [inline contenido/includes/functions.general.php]: Don't replace markup like this, the cHTMLSelectElement clas should handle the setting of the value and the selected state. Trying to replace markup like this should be avoided, this is not a good way.

The generation of the select box is fast, there is no need to complicate the code like this.

**Review by muratpurc — APPROVED** [review]

**Pull request merged**